src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs

```diff
@@ -47,66 +47,102 @@ private static void FreeDeconstructionVariables(ArrayBuilder<DeconstructionVaria
-                    Error(diagnostics, ErrorCode.ERR_DeconstructRequiresExpression, right);
-                }
-
-                if ((object)boundRHS.Type == null)
-                {
-                    // we could still not infer a type for the RHS
-                    FailRemainingInferences(checkedVariables, diagnostics);
-
-                    return new BoundDeconstructionAssignmentOperator(
-                                node, isDeclaration, FlattenDeconstructVariables(checkedVariables), boundRHS,
-                                ImmutableArray<BoundDeconstructionDeconstructStep>.Empty, ImmutableArray<BoundDeconstructionAssignmentStep>.Empty,
-                                voidType, hasErrors: true);
-                }
+                // we could still not infer a type for the RHS
+                FailRemainingInferences(checkedVariables, diagnostics);
+
+                return new BoundDeconstructionAssignmentOperator(
+                            node, isDeclaration, FlattenDeconstructVariables(checkedVariables), boundRHS,
+                            ImmutableArray<BoundDeconstructionDeconstructStep>.Empty,
+                            ImmutableArray<BoundDeconstructionAssignmentStep>.Empty,
+                            ImmutableArray<BoundDeconstructionAssignmentStep>.Empty,
+                            ImmutableArray<BoundDeconstructionConstructionStep>.Empty,
+                            GetSpecialType(SpecialType.System_Void, diagnostics, node),
+                            hasErrors: true);
             }
+
             var deconstructionSteps = ArrayBuilder<BoundDeconstructionDeconstructStep>.GetInstance(1);
+            var conversionSteps = ArrayBuilder<BoundDeconstructionAssignmentStep>.GetInstance(1);
             var assignmentSteps = ArrayBuilder<BoundDeconstructionAssignmentStep>.GetInstance(1);
-            bool hasErrors = !DeconstructIntoSteps(new BoundDeconstructValuePlaceholder(boundRHS.Syntax, boundRHS.Type), node, diagnostics, checkedVariables, deconstructionSteps, assignmentSteps);
+            var constructionStepsOpt = isDeclaration ? null : ArrayBuilder<BoundDeconstructionConstructionStep>.GetInstance(1);
+
+            bool hasErrors = !DeconstructIntoSteps(
+                                    new BoundDeconstructValuePlaceholder(boundRHS.Syntax, boundRHS.Type),
+                                    node,
+                                    diagnostics,
+                                    checkedVariables,
+                                    deconstructionSteps,
+                                    conversionSteps,
+                                    assignmentSteps,
+                                    constructionStepsOpt);
+
+            TypeSymbol returnType = (isDeclaration || hasErrors) ?
+                                        GetSpecialType(SpecialType.System_Void, diagnostics, node) :
+                                        constructionStepsOpt.Last().OutputPlaceholder.Type;
```

Comment:
Does using `void` if `hasErrors` result in additional errors if the result is used? Could we use `constructionStepsOpt.Last().OutputPlaceholder.Type` instead?